src/rulesets/Base/array.jl

```diff
@@ -278,3 +278,132 @@ function rrule(::typeof(fill), value::Any, dims::Int...)
     end
     return fill(value, dims), fill_pullback
 end
+
+#####
+##### `extrema`, `findmax`, `maximum`, etc.
+#####
+
+function rrule(::typeof(extrema), x::AbstractArray{<:Number}; dims=:)
+    ylo, ilo = findmin(x; dims=dims)
+    yhi, ihi = findmax(x; dims=dims)
+    project = ProjectTo(x)
+    extrema_pullback(dys::Tuple{AbstractZero, AbstractZero}) = (NoTangent(), NoTangent())
+    function extrema_pullback((dylo, dyhi))
+        # T = promote_type(eltype(dylo), eltype(dyhi))
+        # @show T  # often Any, when dyhi == NoTangent()
+        T = if dylo isa AbstractZero
+            eltype(dyhi)
+        elseif dyhi isa AbstractZero
+            eltype(dylo)
+        else
+            promote_type(eltype(dylo), eltype(dyhi))
+        end
+        x_thunk = @thunk begin
+            dx = fill!(similar(x, T), false)
```

Comment:
Since you can add AbstractZero to Float64 and get Float64, it might be nice if `promote_type` here would just work, rather than giving `Any`?

Reply:
No, that can't work, as `Zero() * 3.14 isa Zero` while `Zero() + 3.14 isa Float64`. Changed to `T = Base.promote_op(+, typeof(dylo), typeof(dyhi))`.